File: client/dexterity/codegen/dex/instructions/new_order.py
# LOCK-BEGIN[imports]: DON'T MODIFY
from .instruction_tag import InstructionTag
from dataclasses import dataclass
from dexterity.codegen.dex.types import NewOrderParams
from io import BytesIO
from podite import BYTES_CATALOG
from solana.publickey import PublicKey
from solana.transaction import (
    AccountMeta,
    TransactionInstruction,
)
from solmate.utils import to_account_meta
from typing import (
    List,
    Optional,
    Union,
)

# LOCK-END


# LOCK-BEGIN[ix_cls(new_order)]: DON'T MODIFY
@dataclass
class NewOrderIx:
    program_id: PublicKey

    # account metas
    user: AccountMeta
    trader_risk_group: AccountMeta
    market_product_group: AccountMeta
    product: AccountMeta
    aaob_program: AccountMeta
    orderbook: AccountMeta
    market_signer: AccountMeta
    event_queue: AccountMeta
    bids: AccountMeta
    asks: AccountMeta
    system_program: AccountMeta
    fee_model_program: AccountMeta
    fee_model_configuration_acct: AccountMeta
    trader_fee_state_acct: AccountMeta
    fee_output_register: AccountMeta
    risk_engine_program: AccountMeta
    risk_model_configuration_acct: AccountMeta
    risk_output_register: AccountMeta
    trader_risk_state_acct: AccountMeta
    risk_and_fee_signer: AccountMeta
    remaining_accounts: Optional[List[AccountMeta]]

    # data fields
    params: NewOrderParams

    def to_instruction(self):
        keys = []
        keys.append(self.user)
        keys.append(self.trader_risk_group)
        keys.append(self.market_product_group)
        keys.append(self.product)
        keys.append(self.aaob_program)
        keys.append(self.orderbook)
        keys.append(self.market_signer)
        keys.append(self.event_queue)
        keys.append(self.bids)
        keys.append(self.asks)
        keys.append(self.system_program)
        keys.append(self.fee_model_program)
        keys.append(self.fee_model_configuration_acct)
        keys.append(self.trader_fee_state_acct)
        keys.append(self.fee_output_register)
        keys.append(self.risk_engine_program)
        keys.append(self.risk_model_configuration_acct)
        keys.append(self.risk_output_register)
        keys.append(self.trader_risk_state_acct)
        keys.append(self.risk_and_fee_signer)
        if self.remaining_accounts is not None:
            keys.extend(self.remaining_accounts)

        buffer = BytesIO()
        buffer.write(InstructionTag.to_bytes(InstructionTag.NEW_ORDER))
        buffer.write(BYTES_CATALOG.pack(NewOrderParams, self.params))

        return TransactionInstruction(
            keys=keys,
            program_id=self.program_id,
            data=buffer.getvalue(),
        )

# LOCK-END


# LOCK-BEGIN[ix_fn(new_order)]: DON'T MODIFY
def new_order(
    user: Union[str, PublicKey, AccountMeta],
    trader_risk_group: Union[str, PublicKey, AccountMeta],
    market_product_group: Union[str, PublicKey, AccountMeta],
    product: Union[str, PublicKey, AccountMeta],
    aaob_program: Union[str, PublicKey, AccountMeta],
    orderbook: Union[str, PublicKey, AccountMeta],
    market_signer: Union[str, PublicKey, AccountMeta],
    event_queue: Union[str, PublicKey, AccountMeta],
    bids: Union[str, PublicKey, AccountMeta],
    asks: Union[str, PublicKey, AccountMeta],
    fee_model_program: Union[str, PublicKey, AccountMeta],
    fee_model_configuration_acct: Union[str, PublicKey, AccountMeta],
    trader_fee_state_acct: Union[str, PublicKey, AccountMeta],
    fee_output_register: Union[str, PublicKey, AccountMeta],
    risk_engine_program: Union[str, PublicKey, AccountMeta],
    risk_model_configuration_acct: Union[str, PublicKey, AccountMeta],
    risk_output_register: Union[str, PublicKey, AccountMeta],
    trader_risk_state_acct: Union[str, PublicKey, AccountMeta],
    risk_and_fee_signer: Union[str, PublicKey, AccountMeta],
    params: NewOrderParams,
    system_program: Union[str, PublicKey, AccountMeta] = PublicKey("11111111111111111111111111111111"),
    remaining_accounts: Optional[List[AccountMeta]] = None,
    program_id: Optional[PublicKey] = None,
):
    if program_id is None:
        program_id = PublicKey("Dex1111111111111111111111111111111111111111")

    if isinstance(user, (str, PublicKey)):
        user = to_account_meta(
            user,
            is_signer=True,
            is_writable=True,
        )
    if isinstance(trader_risk_group, (str, PublicKey)):
        trader_risk_group = to_account_meta(
            trader_risk_group,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(market_product_group, (str, PublicKey)):
        market_product_group = to_account_meta(
            market_product_group,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(product, (str, PublicKey)):
        product = to_account_meta(
            product,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(aaob_program, (str, PublicKey)):
        aaob_program = to_account_meta(
            aaob_program,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(orderbook, (str, PublicKey)):
        orderbook = to_account_meta(
            orderbook,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(market_signer, (str, PublicKey)):
        market_signer = to_account_meta(
            market_signer,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(event_queue, (str, PublicKey)):
        event_queue = to_account_meta(
            event_queue,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(bids, (str, PublicKey)):
        bids = to_account_meta(
            bids,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(asks, (str, PublicKey)):
        asks = to_account_meta(
            asks,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(system_program, (str, PublicKey)):
        system_program = to_account_meta(
            system_program,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(fee_model_program, (str, PublicKey)):
        fee_model_program = to_account_meta(
            fee_model_program,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(fee_model_configuration_acct, (str, PublicKey)):
        fee_model_configuration_acct = to_account_meta(
            fee_model_configuration_acct,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(trader_fee_state_acct, (str, PublicKey)):
        trader_fee_state_acct = to_account_meta(
            trader_fee_state_acct,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(fee_output_register, (str, PublicKey)):
        fee_output_register = to_account_meta(
            fee_output_register,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(risk_engine_program, (str, PublicKey)):
        risk_engine_program = to_account_meta(
            risk_engine_program,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(risk_model_configuration_acct, (str, PublicKey)):
        risk_model_configuration_acct = to_account_meta(
            risk_model_configuration_acct,
            is_signer=False,
            is_writable=False,
        )
    if isinstance(risk_output_register, (str, PublicKey)):
        risk_output_register = to_account_meta(
            risk_output_register,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(trader_risk_state_acct, (str, PublicKey)):
        trader_risk_state_acct = to_account_meta(
            trader_risk_state_acct,
            is_signer=False,
            is_writable=True,
        )
    if isinstance(risk_and_fee_signer, (str, PublicKey)):
        risk_and_fee_signer = to_account_meta(
            risk_and_fee_signer,
            is_signer=False,
            is_writable=False,
        )

    return NewOrderIx(
        program_id=program_id,
        user=user,
        trader_risk_group=trader_risk_group,
        market_product_group=market_product_group,
        product=product,
        aaob_program=aaob_program,
        orderbook=orderbook,
        market_signer=market_signer,
        event_queue=event_queue,
        bids=bids,
        asks=asks,
        system_program=system_program,
        fee_model_program=fee_model_program,
        fee_model_configuration_acct=fee_model_configuration_acct,
        trader_fee_state_acct=trader_fee_state_acct,
        fee_output_register=fee_output_register,
        risk_engine_program=risk_engine_program,
        risk_model_configuration_acct=risk_model_configuration_acct,
        risk_output_register=risk_output_register,
        trader_risk_state_acct=trader_risk_state_acct,
        risk_and_fee_signer=risk_and_fee_signer,
        remaining_accounts=remaining_accounts,
        params=params,
    ).to_instruction()

# LOCK-END

File: dexteritysdk/src/processor/new_order.rs
use crate::{common::utils::*, sdk_client::SDKClient, KeypairD};
use agnostic_orderbook::state::*;
use anchor_lang::{InstructionData, ToAccountMetas};
use dex::{accounts, instruction, state::enums::OrderType, utils::numeric::Fractional};
use solana_program::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    system_program,
};
use solana_program_test::ProgramTestContext;
use solana_sdk::signature::{Keypair, Signer};

pub fn new_order_ixs(
    aaob_program: Pubkey,
    user: Pubkey,
    trader_risk_group: Pubkey,
    market_product_group: Pubkey,
    product: Pubkey,
    market_signer: Pubkey,
    orderbook: Pubkey,
    event_queue: Pubkey,
    bids: Pubkey,
    asks: Pubkey,
    fee_model_program: Pubkey,
    fee_model_configuration_acct: Pubkey,
    trader_fee_state_acct: Pubkey,
    fee_output_register: Pubkey,
    risk_engine_program: Pubkey,
    risk_model_configuration_acct: Pubkey,
    risk_engine_accounts: &[Pubkey],
    side: Side,
    max_base_qty: Fractional,
    order_type: OrderType,
    self_trade_behavior: SelfTradeBehavior,
    match_limit: u64,
    limit_price: Fractional,
    risk_output_register: Pubkey,
    trader_risk_state_acct: Pubkey,
) -> Vec<Instruction> {
    let params = dex::NewOrderParams {
        side,
        max_base_qty,
        order_type,
        self_trade_behavior,
        match_limit,
        limit_price,
    };
    let (risk_and_fee_signer, _) =
        Pubkey::find_program_address(&[market_product_group.as_ref()], &dex::ID);
    let mut account_metas = accounts::NewOrder {
        aaob_program,
        user,
        trader_risk_group,
        market_product_group,
        product,
        market_signer,
        orderbook,
        event_queue,
        bids,
        asks,
        system_program: system_program::id(),
        fee_model_program,
        fee_model_configuration_acct,
        risk_model_configuration_acct,
        trader_fee_state_acct,
        fee_output_register,
        risk_engine_program,
        risk_output_register,
        trader_risk_state_acct,
        risk_and_fee_signer,
    }
    .to_account_metas(Some(true));
    for key in risk_engine_accounts.iter() {
        account_metas.push(AccountMeta::new(*key, false));
    }
    vec![Instruction {
        program_id: dex::ID,
        data: instruction::NewOrder { params }.data(),
        accounts: account_metas,
    }]
}

pub async fn new_order(
    client: &SDKClient,
    dex_program_id: Pubkey,
    aaob_program_id: Pubkey,
    user: &KeypairD,
    user_trader_risk_group: Pubkey,
    market_product_group: Pubkey,
    product_account: Pubkey,
    market_signer: Pubkey,
    orderbook: Pubkey,
    event_queue: Pubkey,
    bids: Pubkey,
    asks: Pubkey,
    fee_model_program_id: Pubkey,
    fee_model_config_acct: Pubkey,
    trader_fee_state: Pubkey,
    fee_output_register: Pubkey,
    risk_engine_program_id: Pubkey,
    risk_model_configuration_acct: Pubkey,
    risk_engine_accounts: &[Pubkey],
    side: Side,
    max_base_qty: Fractional,
    order_type: OrderType,
    self_trade_behavior: SelfTradeBehavior,
    match_limit: u64,
    limit_price: Fractional,
    out_register_risk_info: Pubkey,
    risk_state_account_info: Pubkey,
) -> SDKResult {
    let ixs = new_order_ixs(
        aaob_program_id,
        user.pubkey(),
        user_trader_risk_group,
        market_product_group,
        product_account,
        market_signer,
        orderbook,
        event_queue,
        bids,
        asks,
        fee_model_program_id,
        fee_model_config_acct,
        trader_fee_state,
        fee_output_register,
        risk_engine_program_id,
        risk_model_configuration_acct,
        risk_engine_accounts,
        side,
        max_base_qty,
        order_type,
        self_trade_behavior,
        match_limit,
        limit_price,
        out_register_risk_info,
        risk_state_account_info,
    );
    client.sign_send_instructions(ixs, vec![user]).await
}
